synapse/rest/client/room.py

```diff
@@ -204,6 +204,11 @@ async def on_PUT(
         if state_key is not None:
             event_dict["state_key"] = state_key
 
+        # Twisted will have processed the args by now.
+        assert request.args is not None
+        if b"ts" in request.args and requester.app_service:
+            event_dict["origin_server_ts"] = parse_integer(request, "ts", 0)
+
```

Comment:
It would be nice to have some tests for this endpoint. Sadly it doesn't look like there are any unit tests for the existing `ts` parameter on `PUT /_matrix/client/r0/rooms/{roomId}/send/{eventType}/{txnId}`. There are a number of tests for both relevant endpoints in general in https://github.com/matrix-org/synapse/blob/a19d01c3d95f5dbd3a4bb181cb70dacd44135a8b/tests/rest/client/test_rooms.py however.

As this is expected to be a feature used across multiple homeserver implementations, it would be good to include tests in https://github.com/matrix-org/complement - the homeserver-agnostic test suite. But if you're not comfortable writing in Golang, then I'd be happy with some Synapse unit tests as well.

Reply:
Ok, I'll work on it. However, for the complement tests: I don't know Golang, so I'll see how quickly I can get into it.